### PR TITLE
Fix array serialization

### PR DESCRIPTION
**Pull request by alexgleason**
Fixes https://github.com/electric-sql/pglite/issues/220

### CODE REVIEW

**Comment by alexgleason** [comment]
@samwillis Is there a better way to run tests than this? This way is slow and quite painful:

```
pnpm build && pnpm exec vitest run -t 'basic esm array params' tests/basic.test.js
```

**Comment by alexgleason** [comment]
I cleaned it up and I think it's a bit better now. Let me know if you have any suggestions!

**Comment by samwillis** [comment]
Thanks for picking this up @alexgleason!

It's the end of the day for me here and so I'll take a proper look in the morning. I don't suppose you could just jot down the strategy you've gone with, from a quick glance (on my phone cooking dinner) I didn't quite follow what fixedParams was doing. Is this influenced by any other JS Postgres's clients?

**Comment by alexgleason** [comment]
@samwillis After sending Parse, we now send a Describe (`S` type) to get parameter info from Postgres.

I loop through parsedParams, and if the oid we inferred with JS code doesn't match the oid Postgres gave us, I re-parse it with the new oid. Then fixedParams gets passed down instead of parsedParams.

File: packages/pglite/src/base.ts [hunk 1]
```diff
@@ -156,47 +156,75 @@ export abstract class BasePGlite
...
-              values: parsedParams.map(([val]) => val),
-            }),
-            options,
-          )),
-          ...(await this.#execProtocolNoSync(
-            serialize.describe({ type: 'P' }),
+        for (const [msg] of await this.#execProtocolNoSync(
+          serialize.parse({
+            text: query,
+            types: params.map((p) => inferType(p)),
```
Comment (alexgleason) [inline packages/pglite/src/base.ts:173]: A part of me wonders, why even bother trying to infer types in JS at all? Postgres knows better. We might as well remove this code that guesses types, and then have the serializers always expect an oid. Then we just serialize everything by oid.

```suggestion
```

File: packages/pglite/src/types.ts [hunk 2]
```diff
@@ -214,6 +227,17 @@ export function serializeType(
...
+/** Try to guess a Postgres type based on the JavaScript data type. */
+export function inferType(x: unknown): number {
+  return (
+    x instanceof Date ? TIMESTAMPTZ :
+    x instanceof Uint8Array ? BYTEA :
+    typeof x === 'boolean' ? BOOL :
+    typeof x === 'bigint' ? INT8 :
+    0
+  )
+}
```
Comment (alexgleason) [inline packages/pglite/src/types.ts:239]: This function was just added to pass "types" into the Parse message, but I think it's pointless.

**Comment by alexgleason** [comment]
I published this branch to https://www.npmjs.com/package/@soapbox.pub/pglite, pulled it into my tests, and would you look at that:

Screenshot from 2024-09-10 15-45-53

Finally able to use pglite in my application.

**Comment by samwillis** [comment]
Hey @alexgleason 

This is great, I really like the strategy of asking PG what the types of the params should be.

I want to spend some time refactoring the type support, and will use this as the base. Plan for early next week.

**Comment by alexgleason** [comment]
I updated this to remove the unnecessary JS type inference code I added. So it's simpler now, and still works perfectly.

I think this is probably acceptable to merge, as there is not much baggage. I would say optionally the function `serializeType` in types.ts could be removed, but it's still used by `serializeArray`, which I think is fine.

@samwillis if you have any specific gripes feel free to give me a scathing review like I'm you're employee. I have time to fix it, and match your preferred code style etc. Since PGlite is also required for my goals (which is my full time job). I just need to get it in a place where everything works right in my project, then it's done. I also need your help with https://github.com/electric-sql/pglite/issues/327 but I think it makes sense to do this one first.

**Comment by alexgleason** [comment]
Thank you very much @samwillis !

**Comment by samwillis** [comment]
Closing in favour of #340 which is based on this but with a larger refactor.

**Pull request closed without merging**